File: librarymanager/librarymanager-be/src/main/java/ru/spring5/elena/librarymanager/services/PublishingHouseService.java
package ru.spring5.elena.librarymanager.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import ru.spring5.elena.librarymanager.api.dto.Edition;
import ru.spring5.elena.librarymanager.api.dto.PublishingHouse;
import ru.spring5.elena.librarymanager.domain.entity.AuthorEntity;
import ru.spring5.elena.librarymanager.domain.entity.EditionEntity;
import ru.spring5.elena.librarymanager.domain.entity.PublishingHouseEntity;
import ru.spring5.elena.librarymanager.domain.repository.EditionRepository;
import ru.spring5.elena.librarymanager.domain.repository.PublishingHouseRepository;
import ru.spring5.elena.librarymanager.mapper.EditionMapper;
import ru.spring5.elena.librarymanager.mapper.PublishingHouseMapper;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;


@RequiredArgsConstructor
@Service
@Slf4j
public class PublishingHouseService {
    private final PublishingHouseRepository publishingHouseRepository;
    private final EditionRepository editionRepository;

    @Transactional(readOnly = true)
    public List<PublishingHouse> findAll() {
        return publishingHouseRepository.findAll()
                .stream()
                .map(PublishingHouseMapper::map)
                .collect(Collectors.toList());
    }
    
    @Transactional(readOnly = true)
    public Page<PublishingHouse> findAllByPage(Pageable pageable) {
        return publishingHouseRepository.findAll(pageable)
                .map(PublishingHouseMapper::map);
    }
    
    @Transactional(readOnly = true)
    public PublishingHouse findPublishingHouseById(Long publishingHouseId) {
        return publishingHouseRepository.findById(publishingHouseId)
                .map(PublishingHouseMapper::map)
                .orElse(null);
    }
    
    @Transactional
    public PublishingHouse save(Long publishingHouseId,  PublishingHouse dto) {
    	return publishingHouseRepository.findById(publishingHouseId)
                .map(entity -> PublishingHouseMapper.map(dto, entity))
                .map(publishingHouseRepository::saveAndFlush)
                .map(PublishingHouseMapper::map)
                .orElse(null);

    }
    
    @Transactional
    public PublishingHouse create(PublishingHouse dto) {
        return   Optional.of(dto)
                .map(PublishingHouseMapper::map)
                .map(publishingHouseRepository::saveAndFlush)
                .map(PublishingHouseMapper::map)
                .orElse(null);

    }
    
    @Transactional(readOnly = true)
    public Page<Edition> findEditionsByPublishingHouseId(Long publishingHouseId, Pageable pageable) {
        return publishingHouseRepository.findById(publishingHouseId)
        	   .map(publHouseEntity -> 
        	   			editionRepository.findAllByPublishingHouse(publHouseEntity, pageable)
        	   	)
               .orElse(Page.empty())
        	   .map(EditionMapper::map);
    }

	public List<Edition> updateEditionsByPublishingHouseId(
			Long publishingHouseId, 
			List<Long> editionIds
	) {
		return publishingHouseRepository
				.findById(publishingHouseId)
				.map(publishingHouseEntity-> {
					publishingHouseEntity.getEditions().clear();
					publishingHouseEntity.getEditions().addAll(
							editionIds.stream()
  							.map(editionRepository::getOne)
  							.collect(Collectors.toList())
  							);
					return publishingHouseEntity;
				 })
				.map(publishingHouseRepository::saveAndFlush)
				.map(PublishingHouseEntity :: getEditions)
				.orElse(Collections.emptySet())
				.stream()
				.map(EditionMapper :: map)
				.collect(Collectors.toList());
	}
    
}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/services/PublishingHouseService.kt
package ru.elenacherev.librarymanager.services

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.elenacherev.librarymanager.api.dto.Edition
import ru.elenacherev.librarymanager.api.dto.PublishingHouse
import ru.elenacherev.librarymanager.domain.entity.EditionEntity
import ru.elenacherev.librarymanager.domain.entity.PublishingHouseEntity
import ru.elenacherev.librarymanager.domain.repository.EditionRepository
import ru.elenacherev.librarymanager.domain.repository.PublishingHouseRepository
import ru.elenacherev.librarymanager.mapper.map
import java.util.*

@Service
class PublishingHouseService (
    val publishingHouseRepository: PublishingHouseRepository,
    val editionRepository: EditionRepository
){
    @Transactional(readOnly = true)
    fun findAll(): List<PublishingHouse> {
        return publishingHouseRepository
            .findAll()
            .map(PublishingHouseEntity::map)
    }

    @Transactional(readOnly = true)
    fun findAllByPage(pageable: Pageable): Page<PublishingHouse>
    {
        return publishingHouseRepository
            .findAll(pageable)
            .map(PublishingHouseEntity::map)
    }

    @Transactional(readOnly = true)
    fun findPublishingHouseById(publishingHouseId: Long): PublishingHouse {
        return publishingHouseRepository
            .findById(publishingHouseId)
            .map(PublishingHouseEntity::map)
            .orElse(null)
    }

    @Transactional
    fun save(publishingHouseId: Long, dto: PublishingHouse): PublishingHouse {
        return publishingHouseRepository
            .findById(publishingHouseId)
            .map(dto::map)
            .map{ publishingHouse: PublishingHouseEntity -> publishingHouseRepository.saveAndFlush(publishingHouse) }
            .map(PublishingHouseEntity::map)
            .orElse(null)
    }

    @Transactional
    fun create(dto: PublishingHouse): PublishingHouse {
        return Optional.of(dto)
            .map(PublishingHouse::map)
            .map{ publishingHouse: PublishingHouseEntity -> publishingHouseRepository.saveAndFlush(publishingHouse) }
            .map(PublishingHouseEntity::map)
            .orElse(null)
    }

    @Transactional(readOnly = true)
    fun findEditionsByPublishingHouseId(publishingHouseId: Long, pageable: Pageable): Page<Edition> {
        return publishingHouseRepository
            .findById(publishingHouseId)
            .map { publHouseEntity -> editionRepository.findAllByPublishingHouse(publHouseEntity, pageable) }
            .orElse(Page.empty())
            .map(EditionEntity::map)
    }

    fun updateEditionsByPublishingHouseId(
        publishingHouseId: Long,
        editionIds: List<Long>
    ): List<Edition> {
        return publishingHouseRepository
            .findById(publishingHouseId)
            .map { publishingHouseEntity ->
                publishingHouseEntity.editions.clear()
                publishingHouseEntity.editions += editionIds.map(editionRepository::getOne)
                return@map publishingHouseEntity
            }
            .map{ publishingHouse: PublishingHouseEntity -> publishingHouseRepository.saveAndFlush(publishingHouse) }
            .map{obj: PublishingHouseEntity -> obj.editions}
            .orElse(mutableSetOf())
            .map(EditionEntity::map)
    }
}